core/bin/external_node/src/config/mod.rs

```diff
@@ -1347,6 +1348,29 @@ impl ExternalNodeConfig {
             tree_component: TreeComponentConfig { api_port: None },
         }
     }
+
+    /// Returns a verified diamond proxy address.
+    /// If local configuration contains the address, it will be checked against the one returned by the main node.
+    /// Otherwise, the remote value will be used. However, using remote value has trust implications for the main
+    /// node so relying on it solely is not recommended.
+    pub fn diamond_proxy_address(&self) -> anyhow::Result<Address> {
+        let remote_diamond_proxy_addr = self.remote.diamond_proxy_addr;
+        let diamond_proxy_addr = if let Some(addr) = self.optional.contracts_diamond_proxy_addr {
+            anyhow::ensure!(
+                addr == remote_diamond_proxy_addr,
+                "Diamond proxy address {addr:?} specified in config doesn't match one returned \
+                by main node ({remote_diamond_proxy_addr:?})"
+            );
```

Comment:
IMO, it could make sense to run this check in `ExternalNodeConfig::fetch_remote()` since this getter may be called multiple times.